enhancements/baremetal/continuous-hw-introspection.md

```diff
@@ -0,0 +1,154 @@
+to report current status of hardware to the user and/or glean metrics about the
+health of the deployment.
+
+At the end of 4.4, a Baremetal IPI deployment does not have continuous hardware introspection
+running on its master or worker nodes. The underlying provisioning service already has the
+capability to perform hardware inspection on each of the servers periodically. During 4.5, the
+goal is to enable continuous hardware introspection all master and worker nodes by running the
+above mentioned service in a DameonSet. Another piece of this solution would be to enable the
+baremetal-operator (BMO) to read this information and update the "status" field of the
+BareMetalHost CRD. Please refer to [1] for the latest BaremetalHost CRD.
+
+## Motivation
+
+This enhancement will allow the user to get continuous updates on all the baremetal
+servers in their Metal3 deployment providing useful information about the hardware
+capabilities of the servers.
+
+### Goals
+
+This enhancement will allow the baremetal host CR associated with each baremetal
+server in a metal3 deployment to contain up-to-date information about its hardware and
+operational status of a server.
+
+### Non-Goals
+
+This proposal does not include any changes to the BareMetalHost CRD.
+
+## Proposal
+
+This enhancement proposes to add a new DaemonSet started by the machine-api-operator
+only for the BareMetal platform. This DaemonSet would run a new container image
+(ironic-hardware-inventory-recorder) on all master and worker nodes. With the new
+addition, it would become possible to have detailed information regarding the
+hardware connected to both master and worker nodes in a metal3 deployment and not
+just when the worker nodes are provisioned. This also allows the user to keep
+track of any changes in hardware capabilities of servers in a metal3 deployment.
+
+Additional details regarding enhancements made to the underlying service required
+for this feature can be found in [3].
+
+### User Stories [optional]
+
+#### Story 1
+
+As an administrator tasked with managing the physical infrastructure, I would like to
+know the current status of the hardware and its capabilities for each of the baremetal
+servers in my Baremetal IPI deployment.
+
+#### Story 2
+
+As an administrator tasked with managing the physical infrastructure, I would like to know
+of any changes in hardware status and capabilities of each baremetal server while it is part
+of my Baremetal IPI deployment.
+
+### Implementation Details/Notes/Constraints [optional]
+
+Continuous hardware introspection involves periodically reading information about baremetal
+server hardware by talking to the server directly. This information includes details
+about devices connected to each server like its CPUs, NICs and storage devices. The
+actual task of hardware introspection is performed by an agent that needs to collect
+information on all master and worker nodes and hence needs to run as a DaemonSet.
```

Comment:
What if there are nodes which are not being driven by metal3? What if I joined my own RHEL7 node to the cluster? What label are you going to have on nodes managed by metal3 which this daemonset will select?

Also what taints do you plan to tolerate with this daemonset?

Reply:
Currently IPI baremetal doesn't support RHEL nodes (and there are no plans to add that AFAIK), but if you manually added a node I suspect it'd run the agent and register with ironic-inspector, but we'd do nothing with that data because there's no corresponding BareMetalHost object - something we'd have to test I guess (although I'm not sure if we'd support joining RHEL nodes to an IPI barmetal cluster?)

We would have to consider this situation if/when automated discovery of nodes is added (currently ironic can do this, but the baremetal-operator doesn't create BareMetalHost objects from discovered nodes).

Reply:
@sadasu  Could we perhaps add a label to all nodes IPI actually deployed, then ensure the DaemonSet only schedules the agent container on those nodes?

Reply:
I believe we support joining RHEL7 (or BYO RHCOS) nodes to other cluster installs. Why would this one be different? The only limitation I'm aware we make (and it's because of code, not support) is that the nodes (RHEL7 or RHCOS) you bring yourself need to be on the same cloud if you configured a cloud provider.   I don't believe we have a cloud provider configured in a cluster installed via bare metal ipi, and thus wouldn't hit the only known restriction.

Reply:
Updated the "Design Details" section with this information.

Reply:
Yes, updated the "Design Details" section with this information.